app-backend/tool/src/main/scala/eval/Interpreter.scala

```diff
@@ -73,6 +73,13 @@ object Interpreter extends LazyLogging {
     case o: Operation => o.args.map(a => overrideParams(a, overrides)).sequence.map(ks => o.withArgs(ks))
   }
 
+  /** Does a given AST have at least one source? */
+  private def hasSources[M: Monoid](ast: MapAlgebraAST): Interpreted[M] = {
```

Comment:
This feels a bit specialized (which is OK for now). As I said above - I think one thing we'll have to consider is how we can make this a more generic thing once we port it out to MAML.